clang/test/SemaTemplate/GH18291.cpp

```diff
@@ -0,0 +1,32 @@
+// RUN: %clang_cc1 -std=c++23 -verify %s
+
+namespace t1 {
+template<bool> struct enable_if { typedef void type; };
+template <class T> class Foo {};
+template <class X> constexpr bool check() { return true; }
+template <class X, class Enable = void> struct Bar {};
+
+template<class X> void func(Bar<X, typename enable_if<check<X>()>::type>) {}
+// expected-note@-1 {{candidate function}}
+
+template<class T> void func(Bar<Foo<T>>) {}
+// expected-note@-1 {{candidate function}}
+
+void g() {
+  func(Bar<Foo<int>>()); // expected-error {{call to 'func' is ambiguous}}
```

Comment:
Interestingly everyone accepts that https://godbolt.org/z/KG9cx4c6f
I think that this is the expected behavior as I am not aware of a disambiguation rule but it's worth double checking @AaronBallman @zygoloid

Reply:
can you add a similar tests where the overload are constrained (both with subsumption, and with either one having a constraint nit satisfied)

Reply:
I am not seeing wording that would allow us to accept this, but it is interesting that we all consistently implement the same rules and accept anyway. Should this be a Core issue in order to avoid breaking working code?

Reply:
Nah, I think this is contrived enough that it's unlikely to affect users.
And if it does we probably want to break them anyway... I'm not even sure what the exact behavior that determines the order is here: Which do you think should be more specialized? 
Breaking code seems like a least worst options that making arbitrary ordering decisions.
In general there are tons of bugs pertaining to default arguments and ordering and we should fix them.
Adding more special cases to partial ordering seem like a bad idea.

If users find issues with the direction we can reconsider or divide a transition strategy (like we did with template template parameter ordering)


And as explain in https://github.com/llvm/llvm-project/issues/18291 - we are not self consistent

Reply:
I don't have a strong opinion, but when 100% of implementations all do the same thing (which we do), it seems rather silly to expect implementations and users to change instead of asking whether the standard should reflect reality.

https://godbolt.org/z/johT7fva8

I don't see a reason to break user code except for conformance, but *nobody conforms* so...

Reply:
Related core issue CWG2160

Reply:
Okay, after playing around a bit more... I don't think I can explain what our behavior is in practice, so maybe this is the correct approach. Certainly the code *seems* ambiguous to me, so I think the changes are defensible. I'm curious if @zygoloid has opinions here, but if not, I think it's fine to move forward.

Reply:
> can you add a similar tests where the overload are constrained (both with subsumption, and with either one having a constraint nit satisfied)

Subsumption is not checked on deduction, that's only used later during overload resolution, which we are not changing, so I am not sure that test would be relevant to this change.

Relatedly, in a callback to another discussion with @zygoloid , we currently check if deduction would produce arguments which would not satisfy the constraints, but according to the specification this should not apply to partial ordering, which we fail to do so before this patch, and this patch introduces another case where we fail to follow the rules.